VERSION-RANGE-SPEC.rst

```diff
@@ -0,0 +1,658 @@
+in this document.
+
+
+Version range specifier
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A version range specifier (aka. "vers") is a URI string using the ``vers``
+URI-scheme with this syntax::
+
+   vers:<versioning-scheme>/<version-constraint>|<version-constraint,version-constraint>,...
+
+For example to define a set of versions that contains either version ``1.2.3``,
+or any versions greater than or equal to ``2.0.0`` but less than ``5.0.0`` using
+the ``node-semver`` versioning scheme, the version range specifier will be::
+
+    vers:npm/1.2.3|>=2.0.0,<5.0.0
+
+Each ``<version-constraint>`` in the pipe-separated list is either a simple
+constraint such as::
+
+    <comparator:version>
+
+Or a composite constraint grouping multiple ``<version-constraint>`` joined by
+a comma such as::
+
+    <comparator:version>,<comparator:version>...
+
+The pipe is a logical `OR` and the comma is a logical `AND`.
+
+A version range specifier is therefore an "OR of ANDs" where there are two
+levels of constraints that a version should satisfy to be part of the range:
+
+- At the first level, anyone of the constraints should be satisfied
+- At the second level, all of the constraints must be satisfied
+
+This is also called a "disjunctive normal form" in boolean logic.
+See https://en.wikipedia.org/wiki/Disjunctive_normal_form for details.
+
+``vers`` is the URI-scheme and is an acronym for "VErsion Range Specifier". It
+has been selected because it is short, obviously about version and available
+for a future formal registration for this URI-scheme at the IANA registry.
+
+
+``<versioning-scheme>``
+------------------------
+
+The ``<versioning-scheme>`` (such as ``npm``, ``deb``, etc.) determines:
+
+- the specific notation and conventions used for a version string encoded in
+  this scheme. Versioning schemes often specify a version segments separator and
+  the meaning of each version segments, such as [major.minor.patch] in semver.
+
+- how two versions are compared as greater or lesser to determine if a version
+  is within or outside a range.
+
+- how a versioning scheme-specific range notation can be transformed in the
+  ``vers`` simplified notation defined here.
+
+- by convention the versioning scheme should be the same string as the Package
+  URL package type for a given package ecosystem. It is OK to have other schemes
+  beyond the purl type and even schemes that are specific to a single package.
+
+The ``<versioning-scheme>`` is followed by a slash "/".
+
+
+``<version-constraint>``
+----------------------------
+
+After the ``<versioning-scheme>`` and "/" there are one or more
+``<version-constraint>`` separated by a pipe "|". The pipe "|" means that
+**any** of these constraints must be satisfied for a version to be resolved as
+within this version range.
+
+Each  ``<version-constraint>`` of this pipe-separated list can be either a
+single constraint or a list of constraints separated in turn by an comma "," as
+in ``1.2.3|>=2.0.0,<5.0.0``.
```

Comment:
May I suggest a more OSV/CVE 5.0 like approach here? OSV used to have a similar approach to this (using a disjunctive normal form), but we realised there was a much more generalisable and simpler way to this as we were discussing this in depth with the CVE WG.

This would constrain the way these can be specified a little, leading to simpler evaluation, less potential human error and with no loss of expressiveness.

The additional restrictions would be:

- Only `>=` and `<`, or `=/no operator` are allowed. (The first two operators would correspond to ["introduced" and "fixed" in OSV](https://ossf.github.io/osv-schema/#examples)).
- Only a single constraint per pipe ("|") item. The meaning of pipe ("|") would also change, in that they just become a separator for the different constraints / points.

I.e. instead of `1.2.3|>=2.0.0,<5.0`, we just have `1.2.3|>=2.0.0|<5.0`.

To evaluate them, these essentially become sign posts / events on a version "timeline" (which could instead be a tree because of branching). The algorithm to evaluate this would be almost identical to how CVE 5.0 / OSV does it:

```
# e.g. v == '2.0.0'
# constraints = ['1.2.3', '>=2.0.0', '<5.0.0']

func is_v_affected(v, constraints):
  status = unaffected
  sorted_constraints = sorted(constraints)
  for constraint in sorted_constraints:
    if is_equals_operator(constraint) and constraint == v:
      return True

    if v >= constraint and is_greater_equals(constraint):
      status = affected

    if v >= constraint and is_less_than(constraing):
      status = unaffected
     
 return status
```

Why would we want this instead?

1. A more restrictive way of specifying ranges means more consistency and less chance of human error when writing these ranges (e.g. there's no way to write overlapping ranges). They're also much easier to write evaluators for. This should still be just as expressive.
2. `<version` entries have the side effect of pointing to versions that contain fixes. So, they are more descriptive to users who want to know what version to update to.
3. This also generalizes well to specifying affected parts of [complicated git commit trees](https://ossf.github.io/osv-schema/#examples) that are not possible to express using a disjunctive normal form with the same semantics.

@pombredanne WDYT?

Reply:
@oliverchang re: WDYT?

I think this is effing **brilliant** and much better... I cannot find a case that would not fit for vulnerabilities or dependencies. This effectively becomes much simpler to implement in the general case. It also provides a natural "canonical" sort order, which is simply that of the versions.  It may make the range conversion code of some range notation more complex in some complex edge cases: complex code for complex cases is perfectly OK in my book.

There is only one corner case where I wonder if adding a `!=` (not equal operator) could help: 

- I can express this as an interval by adding a gap in an interval ... eg., if in `>=2 | <6` I want to exclude version `5`, then I can rewrite this `>=2|<6` as `>=2 | <5 | >=5.1 | <6`. 
- This may demand to know what is the "smallest" next version after `5` and may not be strictly equivalent to `!=5`?
- I reckon this is **not** a use case that could happen for vulnerabilities and bugs, yet is somewhat common for dependencies as in: `use any version 2 or later, up to and excluding version 6, except for version 5 that I know is buggy`. 
- I am split as  `>=2 |  !=5.1 | <6` may feel a bit simpler ... yet these case may not be exceptional enough to warrant their own extra operator... e.g. less is more,  and `>=` and `<`, or `=`/no operators may be enough?

Reply:
@pombredanne I'm very glad you like this! 

>  It may make the range conversion code of some range notation more complex in some complex edge cases: complex code for complex cases is perfectly OK in my book.

Cannot agree more :) 

> There is only one corner case where I wonder if adding a != (not equal operator) could help:

I agree that less is more! It's a corner case that can still be expressed using `>=`, `<` and `=` alone. I would expect most ranges to represent different release branches (e.g. `>= 2 | < 2.2 | >= 3 | < 3.0.3 | >= 6 | < 6.1`), where the gaps would be naturally encoded by omission. 

It's also easier to add operators later if it turns out != is really needed in a lot of cases, rather than removing it.

Reply:
> It's also easier to add operators later if it turns out != is really needed in a lot of cases, rather than removing it.

That's an excellent point.

Reply:
@oliverchang You feedback is mucho welcomed on the latest push. Thank you ++ for this.